cmd/sink/event_data.go

```diff
@@ -0,0 +1,138 @@
+	Kind       string
+	Name       string
+	Namespace  string
+}
+
+// implement the Stringers interface so that golang can convert ArchiveEntries to string
+func (entry *ArchiveEntry) String() string {
+	return string(entry.Event.Data())
+}
+
+// the timestamp when the resource was created
+func (entry *ArchiveEntry) Created() string {
+	return entry.Data.Created
+}
+
+// the timestamp when the resource was last updated
+func (entry *ArchiveEntry) LastUpdated() string {
+	return entry.Data.LastUpdated
+}
+
+// the version of the resources
+func (entry *ArchiveEntry) ResourceVersion() string {
+	return entry.Data.Metadata.ResourceVersion
+}
+
+// checks that the cloudevents has the appropriate extensions set and have values that are the right type. Additionally
+// it checks that the cloudevent's data has the necessary fields. If all conditions are not met, it returns an error
+func NewArchiveEntry(event cloudevents.Event) (*ArchiveEntry, error) {
+	eventExtensions := event.Extensions()
+	apiVersion, err := types.ToString(eventExtensions[apiVersionExtension])
+	if err != nil {
+		return nil, err
+	}
+	kind, err := types.ToString(eventExtensions[kindExtension])
+	if err != nil {
+		return nil, err
+	}
+	name, err := types.ToString(eventExtensions[nameExtension])
+	if err != nil {
+		return nil, err
+	}
+	namespace, err := types.ToString(eventExtensions[namespaceExtension])
+	if err != nil {
+		return nil, err
+	}
+	payload := ResourceData{}
+	err = event.DataAs(&payload)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ArchiveEntry{
+			Event:      event,
+			Data:       payload,
+			ApiVersion: apiVersion,
+			Kind:       kind,
+			Name:       name,
+			Namespace:  namespace,
+		},
+		nil
+}
+
+// Writes the kubernetes resource held by entry to the database using the provided connection string. Timeouts and
+// deadlines for writing to the database are handled by ctx. If the database transaction fails, it will return an
+// error and attempt to rollback the transaction
+func (entry ArchiveEntry) WriteToDatabase(ctx context.Context, dbConn *sql.DB) error {
+	tx, err := dbConn.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("could not begin transaction: %s", err)
+	}
+	_, execErr := tx.ExecContext(ctx,
+		writeQuery,
+		entry.ApiVersion,
+		entry.Kind,
+		entry.Name,
+		entry.Namespace,
+		entry.ResourceVersion(),
+		entry.String(),
+		entry.Created(),
+		entry.LastUpdated(),
+	)
+	if execErr != nil {
+		rollbackErr := tx.Rollback()
+		if rollbackErr != nil {
+			return fmt.Errorf("write to database failed: %s and unable to rollback transaction: %s", execErr, rollbackErr)
+		}
+		return fmt.Errorf("write to database failed: %s", execErr)
+	}
+	execErr = tx.Commit()
+	if execErr != nil {
+		rollbackErr := tx.Rollback()
+		if rollbackErr != nil {
+			return fmt.Errorf("write to database failed: %s and unable to rollback transaction: %s", execErr, rollbackErr)
+		}
+		return fmt.Errorf("write to database failed: %s", execErr)
+	}
```

Comment:
The label in these error messages is the same as above, which means when they appear in the log you will not be able to tell which spot in the code failed. I suggest updating the text label to make the action clear.

```suggestion
	execErr = tx.Commit()
	if execErr != nil {
		rollbackErr := tx.Rollback()
		if rollbackErr != nil {
			return fmt.Errorf("commit to database failed: %s and unable to rollback transaction: %s", execErr, rollbackErr)
		}
		return fmt.Errorf("commit to database failed: %s", execErr)
	}
```

Reply:
fixed